### PR TITLE
Make rclcpp::Duration support scale operation

### PR DESCRIPTION
**Pull request by jwang11**
Duration scale is a convinient operation which had supported in ROS.
This commit make ROS2 support it.

Signed-off-by: jwang <jing.j.wang@intel.com>

### CODE REVIEW

File: rclcpp/src/rclcpp/duration.cpp [hunk 1]
```diff
@@ -182,6 +182,12 @@ Duration::operator-(const rclcpp::Duration & rhs) const
     rcl_duration_.nanoseconds - rhs.rcl_duration_.nanoseconds);
 }
 
+Duration
+Duration::operator*(double scale) const
+{
+  return Duration(rcl_duration_.nanoseconds * scale);
```
Comment (tfoote) [inline rclcpp/src/rclcpp/duration.cpp]: Please add underflow and overflow checking here. We are checking + and - operations we should check * too.

File: rclcpp/test/test_duration.cpp [hunk 2]
```diff
@@ -54,6 +54,10 @@ TEST(TestDuration, operators) {
   EXPECT_EQ(sub.nanoseconds(), (rcl_duration_value_t)(young.nanoseconds() - old.nanoseconds()));
   EXPECT_EQ(sub, young - old);
 
+  rclcpp::Duration scale = old * 3;
+  EXPECT_EQ(scale.nanoseconds(), (rcl_duration_value_t)(old.nanoseconds() * 3));
+  EXPECT_EQ(scale, old * 3);
```
Comment (dirk-thomas) [inline rclcpp/test/test_duration.cpp]: This line doesn't seem to test anything. `scale` is defined just as `old * 3` two lines above. So this will always be `true` if the operator is deterministic.

Comment (jwang11) [reply rclcpp/test/test_duration.cpp]: I know it, the reason I add this line is that the tests for operation +/- both include it, see
'''
  rclcpp::Duration add = old + young;
  EXPECT_EQ(add.nanoseconds(), (rcl_duration_value_t)(old.nanoseconds() + young.nanoseconds()));
  EXPECT_EQ(add, old + young);

  rclcpp::Duration sub = young - old;
  EXPECT_EQ(sub.nanoseconds(), (rcl_duration_value_t)(young.nanoseconds() - old.nanoseconds()));
  EXPECT_EQ(sub, young - old);
'''

**Comment by mikaelarguedas** [comment]
Thanks @jwang11 for the contribution.
@tfoote of @dirk-thomas to run CI and merge if green

File: rclcpp/src/rclcpp/duration.cpp [hunk 3]
```diff
@@ -182,6 +182,36 @@ Duration::operator-(const rclcpp::Duration & rhs) const
     rcl_duration_.nanoseconds - rhs.rcl_duration_.nanoseconds);
 }
 
+void
+bounds_check_duration_scale(int64_t dns, double scale, uint64_t max)
+{
+  auto abs_dns = (uint64_t)std::abs(dns);
+  auto abs_scale = (uint64_t)std::abs(scale);
```
Comment (tfoote) [inline rclcpp/src/rclcpp/duration.cpp]: Casting this to be an unsigned integer I think is going to do the wrong thing here. It's used lower as if it's still a float.

File: rclcpp/src/rclcpp/duration.cpp [hunk 4]
```diff
@@ -182,6 +182,36 @@ Duration::operator-(const rclcpp::Duration & rhs) const
...
+
+  if ((abs_scale >= 1.0 && abs_dns > (uint64_t)(max / abs_scale)) ||
+    (abs_scale < 1.0 && (uint64_t)(abs_dns * abs_scale > max)))
```
Comment (tfoote) [inline rclcpp/src/rclcpp/duration.cpp]: If we're below an absolute scaling factor of 1 the value can only get smaller which will never overflow or underflow a signed integer. I think this whole 2nd case can be dropped.

File: rclcpp/src/rclcpp/duration.cpp [hunk 5]
```diff
@@ -182,6 +182,36 @@ Duration::operator-(const rclcpp::Duration & rhs) const
...
+  {
+    if ((dns > 0 && scale > 0) || (dns < 0 && scale < 0)) {
+      throw std::overflow_error("addition leads to int64_t overflow");
```
Comment (tfoote) [inline rclcpp/src/rclcpp/duration.cpp]: This should be 'duration scaling' not 'addition'

File: rclcpp/src/rclcpp/duration.cpp [hunk 6]
```diff
@@ -182,6 +182,36 @@ Duration::operator-(const rclcpp::Duration & rhs) const
...
+    } else {
+      throw std::underflow_error("addition leads to int64_t underflow");
```
Comment (tfoote) [inline rclcpp/src/rclcpp/duration.cpp]: This should be 'duration scaling' not 'addition'

File: rclcpp/test/test_duration.cpp [hunk 7]
```diff
@@ -54,6 +54,9 @@ TEST(TestDuration, operators) {
   EXPECT_EQ(sub.nanoseconds(), (rcl_duration_value_t)(young.nanoseconds() - old.nanoseconds()));
   EXPECT_EQ(sub, young - old);
 
+  rclcpp::Duration scale = old * 3;
+  EXPECT_EQ(scale.nanoseconds(), (rcl_duration_value_t)(old.nanoseconds() * 3));
```
Comment (tfoote) [inline rclcpp/test/test_duration.cpp]: This should also have tests for the underflow and overflow. In particular I'd like to see at least all 4 cases of underflow and overflow with positive and negative scaling factors.

**Comment by dirk-thomas** [comment]
Please make sure to run the unit tests of the package locally. The linters will e.g. enforce to use C++ cast rather than C casts.

File: rclcpp/test/test_duration.cpp (hunk 7 above)
Comment (jwang11) [reply rclcpp/test/test_duration.cpp]: Thanks for comments, they are all reasonable.

**Comment by tfoote** [comment]
CI: 
Linux: [Build Status](https://ci.ros2.org/job/ci_linux/3969/)
Linux arm64: [Build Status](https://ci.ros2.org/job/ci_linux-aarch64/1082/)
OSX: [Build Status](https://ci.ros2.org/job/ci_osx/3306/)
Windows: [Build Status](https://ci.ros2.org/job/ci_windows/4069/)

File: rclcpp/src/rclcpp/duration.cpp [hunk 8]
```diff
@@ -182,6 +182,34 @@ Duration::operator-(const rclcpp::Duration & rhs) const
     rcl_duration_.nanoseconds - rhs.rcl_duration_.nanoseconds);
 }
 
+void
+bounds_check_duration_scale(int64_t dns, double scale, uint64_t max)
+{
+  auto abs_dns = static_cast<uint64_t>(std::abs(dns));
+  auto abs_scale = std::abs(scale);
```
Comment (tfoote) [inline rclcpp/src/rclcpp/duration.cpp]: This is apparently ambiguous on OSX: https://ci.ros2.org/job/ci_osx/3306/consoleFull

```
03:04:22 [ 38%] Building CXX object CMakeFiles/rclcpp.dir/src/rclcpp/contexts/default_context.cpp.o
03:04:23 [ 39%] Building CXX object CMakeFiles/rclcpp.dir/src/rclcpp/duration.cpp.o
03:04:23 /Users/osrf/jenkins/workspace/ci_osx/ws/src/ros2/rclcpp/rclcpp/src/rclcpp/duration.cpp:189:20: error: call to 'abs' is ambiguous
03:04:23   auto abs_scale = std::abs(scale);
03:04:23                    ^~~~~~~~
03:04:23 /usr/include/stdlib.h:137:6: note: candidate function
03:04:23 int      abs(int) __pure2;
03:04:23          ^
03:04:23 /Library/Developer/CommandLineTools/usr/include/c++/v1/stdlib.h:115:44: note: candidate function
03:04:23 inline _LIBCPP_INLINE_VISIBILITY long      abs(     long __x) _NOEXCEPT {return  labs(__x);}
03:04:23                                            ^
03:04:23 /Library/Developer/CommandLineTools/usr/include/c++/v1/stdlib.h:117:44: note: candidate function
03:04:23 inline _LIBCPP_INLINE_VISIBILITY long long abs(long long __x) _NOEXCEPT {return llabs(__x);}
03:04:23                                            ^
03:04:23 /Users/osrf/jenkins/workspace/ci_osx/ws/src/ros2/rclcpp/rclcpp/src/rclcpp/duration.cpp:203:13: error: no member named 'isnormal' in namespace 'std'
03:04:23   if (!std::isnormal(scale)) {
03:04:23        ~~~~~^
03:04:23 2 errors generated.
03:04:23 make[2]: *** [CMakeFiles/rclcpp.dir/build.make:212: CMakeFiles/rclcpp.dir/src/rclcpp/duration.cpp.o] Error 1
03:04:23 make[1]: *** [CMakeFiles/Makefile2:855: CMakeFiles/rclcpp.dir/all] Error 2
03:04:23 make: *** [Makefile:141: all] Error 2
03:04:23 
```

Comment (wjwwood) [reply rclcpp/src/rclcpp/duration.cpp]: You want `std::fabs`: http://en.cppreference.com/w/cpp/numeric/math/fabs

Comment (wjwwood) [reply rclcpp/src/rclcpp/duration.cpp]: Though, it should probably work anyways...

Comment (wjwwood) [reply rclcpp/src/rclcpp/duration.cpp]: You might just need to `#include <cmath>`. The standard seems to have been inconsistent on that in the past, which would allow for two different implementations to give you different results.

Comment (jwang11) [reply rclcpp/src/rclcpp/duration.cpp]: Ok, I'll explicitly add #include <cmath>, thank you

**Comment by dhood** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/3976/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/1089/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/3313/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/4078/)

File: rclcpp/src/rclcpp/duration.cpp [hunk 9]
```diff
@@ -182,6 +183,34 @@ Duration::operator-(const rclcpp::Duration & rhs) const
...
+Duration::operator*(double scale) const
+{
+  if (!std::isnormal(scale)) {
+    throw std::runtime_error("abnormal scale in rclcpp::Duration");
+  }
+  bounds_check_duration_scale(
+    this->rcl_duration_.nanoseconds,
+    scale,
+    std::numeric_limits<rcl_duration_value_t>::max());
+  return Duration(rcl_duration_.nanoseconds * scale);
```
Comment (dhood) [inline rclcpp/src/rclcpp/duration.cpp]: this is giving a warning on windows: http://ci.ros2.org/job/ci_windows/4078/
```
'argument': conversion from 'double' to 'rcl_duration_value_t', possible loss of data
```

Comment (jwang11) [reply rclcpp/src/rclcpp/duration.cpp]: I guess a explicit cast can avoid this warning, however locally I have no windows build environment, please help re-trigger CI to check.

**Comment by sloretz** [comment]
CI with 3dacb27

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/3981/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/1097/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/3318/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/4086/)

**Comment by jwang11** [comment]
Windows warning disappeared. From log, the failure on macOS and Windows are python test script to execute command, which should not relate to this PR.

File: rclcpp/src/rclcpp/duration.cpp [hunk 10]
```diff
@@ -182,6 +183,34 @@ Duration::operator-(const rclcpp::Duration & rhs) const
...
+    } else {
+      throw std::underflow_error("duration scaling leads to int64_t underflow");
+    }
+  }
+}
+
+Duration
+Duration::operator*(double scale) const
+{
+  if (!std::isnormal(scale)) {
```
Comment (dhood) [inline rclcpp/src/rclcpp/duration.cpp]: @tfoote is it appropriate to prevent users from having a scale of 0 here? Seems too strict

Comment (tfoote) [reply rclcpp/src/rclcpp/duration.cpp]: Yeah, a zero scale value would be valid. I think `isfinite` would be a better check.

Comment (jwang11) [reply rclcpp/src/rclcpp/duration.cpp]: I agree, in theory zero is valid to scale although this operation is meaningless. Ok, I'll update code from isnormal to isfinite which allow both normal and zero.

**Comment by jwang11** [comment]
If this PR is still under review by someone? Actually I have a porting work from ROS1 to ROS2 which reply on this patch, is it possible expedite the process? thanks

**Comment by tfoote** [comment]
@jwang11  It's good to comment back when the existing reviewers comments are addressed so we know it's ready to look at again.

I've restarted CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/4087/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/1182/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/3382/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/4147/)

**Comment by tfoote** [comment]
Unfortunately the branch was now out of date. I rebased it to https://github.com/ros2/rclcpp/tree/duration_scale

And started CI again.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/4090/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/1185/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/3384/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/4149/)

**Comment by tfoote** [comment]
The only failures are composition 2 on osx and 4 on windows: https://github.com/ros2/build_cop/issues/86

**Comment by tfoote** [comment]
Squashed rebased and merged in https://github.com/ros2/rclcpp/commit/af6e86c522d674a58f935cabceea2108c7f64589
